### DISCUSSION TITLE
Some tables reads fine and some not, what is the reason?

### DISCUSSION THREAD

**Discussion by dakanadaka** [simolus3/drift#1862] [General]
Hello there,
see also code: https://gist.github.com/dakanadaka/a03684484f208e8ac860d71959c4b004
I have old dbisam datbase and I can connect with python. With pyton I made the script that reads DBISAM table and put all tables and data to sqlite database.

Now I have one problem in flutter and drift. Some of tables drift wants to read and some not. In dbeaver I can see them all. 
In the flutter application with drift I can see only some of them. 

lets say:

one table can be retrieved and some not. I don't get that.

This prep.... cant be seen in flutter and other dis... can be seen
![dis](https://user-images.githubusercontent.com/4201758/170832825-cf3d8125-6e61-4364-b903-51f4f71a66aa.PNG)
![prep](https://user-images.githubusercontent.com/4201758/170832778-32921862-f3d3-40df-ac36-02193b96295d.PNG)

 Here is the code:
 ```dart
import '../models/bom.dart';
//import 'package:flutter/services.dart' show rootBundle;

part 'database.g.dart';
@DriftDatabase(tables: [
    User,
    Issled,
    Pacient,
    Flutter,
    Bom,
    Afterterapia,
    Disease,
    Preparat,
])
class AppDatabase extends _$AppDatabase{
  AppDatabase() : super(_openConnection());
  @override
  int get schemaVersion => 1;

  Future<List<PacientData>> getAllPacients() async {
    return await pacient.select().get();
  }

  //GET PACIENT LIST
  Future<List<DiseaseData>> getAllDisease() async {
    return await disease.select().get();
  }

  //GET PACIENT LIST
  Future<List<PreparatData>> getAllPreparat() async {
    return await preparat.select().get();
  }
}
```

I delete database.g.dart and un the command: flutter pub run build_runner build
but no luck. 

Here is the list of tables that run well:
User, Flutter, Bom, Disease.

other are not working

main.dart page
```dart
import 'package:fluent_ui/fluent_ui.dart';
import 'package:provider/provider.dart';
import 'package:win2022/database/database.dart';
import 'package:win2022/providers/scan_provider.dart';
import 'package:win2022/providers/user_provider.dart';
import 'screens/home_page.dart';
void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(
    MultiProvider(providers: [
      ChangeNotifierProvider(create: (_) => UserProvider()),
      ChangeNotifierProvider(create: (_) => ScanProvider()),
      Provider<AppDatabase>(create: (context) => AppDatabase(),)
    ],
      child: MyApp(),),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  
  @override
  Widget build(BuildContext context) {
    return FluentApp(
        debugShowCheckedModeBanner: false,
        title: 'Flutter Demo',
        theme: ThemeData(
          accentColor: Colors.blue,
        ),
        home: const HomePage(),
      );
  }
}
```

**Comment by dakanadaka** [comment]
I just checked disease table and it is not big.
but preparat is like 100.000 records, is there a limit to add to sqlite query?

**Reply by dakanadaka** [reply]
tried: return await (select(preparat)..limit(10, offset: 15)).get(); and didn't work

**Comment by simolus3** [comment]
Could you post the error you get when you try to load some of the failing tables?

I believe this might be because drift expects a slightly different column name than what you have in the database, so the mapping from the database row to Dart fails. But to say for sure, I need the `CREATE TABLE` statement for the offending table, the Dart table class and the exact exception and stack trace.

**Comment by dakanadaka** [comment]
There is no error, except that there is no data.
![prep2](https://user-images.githubusercontent.com/4201758/170844683-50a10b87-9d1b-48ac-b868-101be92fcdac.PNG)

and code:
`import 'package:drift/drift.dart';

class Preparat extends Table{
  IntColumn get ind => integer().autoIncrement()();
  IntColumn get indgroup => integer()();
  TextColumn get name => text()();
  IntColumn get orggroup => integer()();
  TextColumn get sex => text()();
  TextColumn get fname => text()();
  IntColumn get oldind => integer()();
  RealColumn  get min_level_man_blood => real()();
  RealColumn  get max_level_man_blood => real()();
  RealColumn  get min_level_women_blood => real()();
  RealColumn  get max_level_women_blood => real()();
  IntColumn get hilda_freq_min => integer()();
  IntColumn get hilda_freq_max => integer()();
  TextColumn get unit_blood => text()();
  RealColumn  get value_blood => real()();
  TextColumn get spc_code => text()();

}
`

**Reply by dakanadaka** [reply]
![prep3](https://user-images.githubusercontent.com/4201758/170844725-f723b768-cd8c-4c05-88bb-1b4738982ce3.PNG)

**Comment by dakanadaka** [comment]
`CREATE TABLE IF NOT EXISTS 'preparat' ( 'ind' INTEGER PRIMARY KEY AUTOINCREMENT, 'indgroup' INTEGER, 'name' TEXT, 'orggroup' INTEGER, 'sex' TEXT, 'fname' TEXT, 'oldind' INTEGER, 'min_level_man_blood' REAL, 'max_level_man_blood' REAL, 'min_level_woman_blood' REAL, 'max_level_woman_blood' REAL, 'hilda_freq_min' INTEGER, 'hilda_freq_max' INTEGER, 'unit_blood' TEXT, 'value_blood' REAL, 'spc_code' TEXT);
`

From python to make table

**Reply by dakanadaka** [reply]
From python I can get all records lik e this:
`import db
import vec_tables
from datetime import date
import sys


conn    = db.DB()
odbc    = conn.conn_odbc()
sqlite  = conn.conn_sqlite()


sqlite_check_result_exists = "SELECT * from preparat WHERE ind = 2 ;"
print(sqlite_check_result_exists)
sqlite_cur = sqlite.cursor()
sqlite_cur.execute(sqlite_check_result_exists)
rows = sqlite_cur.fetchall()
print(rows)
for row in rows:
    print(row)


sqlite_check_result_exists = "SELECT * from preparat;"
print(sqlite_check_result_exists)
sqlite_cur = sqlite.cursor()
sqlite_cur.execute(sqlite_check_result_exists)
rows = sqlite_cur.fetchall()
print(rows)
for row in rows:
    print(row)

sqlite_check_result_exists = "SELECT count(*) from preparat;"
print(sqlite_check_result_exists)
sqlite_cur = sqlite.cursor()
sqlite_cur.execute(sqlite_check_result_exists)
rows = sqlite_cur.fetchall()
print(rows)
for row in rows:
    print(row)`

There are 133280 records in this table.

**Comment by dakanadaka** [comment]
how to debug this? Any ideas?

**Reply by simolus3** [reply]
You could try enabling `logStatements` on the `NativeDatabase` constructor. That will log the queries that drift is issuing, which may point to what's going wrong if the query itself looks off. You can also try to set a breakpoint in the generated  `<YourDataClass>.fromData` factory which gets called when drift maps a row back to Dart - perhaps something is being dropped there?

**Reply by dakanadaka** [reply]
"NativeDatabase" thanks for the tip, now I can see all the queries, but still no luck
"SELECT * FROM preparat LIMIT 10 OFFSET 15;" dont get anything and in dbweaver I get the results.

"<YourDataClass>.fromData factory which gets called when drift maps a row back to Dart" -> will try

**Reply by simolus3** [reply]
FWIW, if you're not getting any results, it might also be that you're not importing the database correctly? Drift will create the tables if the database file doesn't exist. So if the import of the old database doesn't go through you'd also see this behavior.

**Comment by North101** [comment]
Might also be worth checking that the created db is what drift expects it to be by running:
```
await db.validateDatabaseSchema();
```

**Reply by dakanadaka** [reply]
How  to call db.validateDatabaseSchema()?

in database.dart? https://gist.github.com/dakanadaka/a03684484f208e8ac860d71959c4b004#file-database-dart

**Reply by dakanadaka** [reply]
@simolus3 can you help me with this? Do you have any example. how to use on db object?

**Reply by simolus3** [reply]
@dakanadaka You need to import an API from `drift_dev` as well, see [this section of the documentation](https://drift.simonbinder.eu/docs/advanced-features/migrations/#verifying-a-database-schema-at-runtime) for an example.

**Comment by dakanadaka** [comment]
oke, it was null safety. If database has records that are None then it doesn't work.

**Reply by simolus3** [reply]
Drift's columns are non-nullable by default yeah. You can add `.nullable()` to the column's definition to allow null values (and you'll also re-generate sources after that).
